### PR TITLE
Fix composite input handling

### LINKED ISSUE #1319: Composite action inputs clobbered
### Bug report info

```plain text
$ act --bug-report
act version:            0.2.30
GOOS:                   linux
GOARCH:                 amd64
NumCPU:                 8
Docker host:            DOCKER_HOST environment variable is unset/empty.
Sockets found:
        /var/run/docker.sock
Config files:
        /home/gwarner/.actrc:
                -P ubuntu-latest=catthehacker/ubuntu:act-latest
                -P ubuntu-20.04=catthehacker/ubuntu:act-20.04
                -P ubuntu-18.04=catthehacker/ubuntu:act-18.04
                ubuntu-16.04=catthehacker/ubuntu:act-16.04
Build info:
        Go version:            go1.18.5
        Module path:           github.com/nektos/act
        Main version:          (devel)
        Main path:             github.com/nektos/act
        Main checksum:
        Build settings:
                -compiler:            gc
                CGO_ENABLED:          1
                GOARCH:               amd64
                GOOS:                 linux
                GOAMD64:              v1
Docker Engine:
        Engine version:        20.10.17
        Engine runtime:        runc
        Cgroup version:        2
        Cgroup driver:         systemd
        Storage driver:        overlay2
        Registry URI:          https://index.docker.io/v1/
        OS:                    NixOS 22.05 (Quokka)
        OS type:               linux
        OS version:            22.05
        OS arch:               x86_64
        OS kernel:             5.15.60
        OS CPU:                8
        OS memory:             63521 MB
        Security options:
                name=seccomp,profile=default
                name=cgroupns
```

### Command used with act

```sh
act -P ubuntu-18.04=nektos/act-environments-ubuntu:18.04 -W .github/workflows/pipeline.yaml
```

### Describe issue

In action, `${{ inputs... }}` not available after 'using' actions/setup-node or actions/setup-java.

### Link to GitHub repository

https://github.com/gdw2/act-test

### Workflow content

```yml
name: Pipeline
on:
  push:

jobs:
  test-orders:
    runs-on: [ viarise ]
    steps:
      - uses: actions/checkout@v2
      - uses: ./.github/actions/test
        with:
          module: some-module


--- (action below) ---

name: Test Action
inputs:
  module:
    required: true
    description: some description

runs:
  using: composite
  steps:
    - shell: bash
      run: |
        echo TEST1 ${{ inputs.module }}
    - uses: actions/setup-node@v2
      with:
        node-version: '16'
    - shell: bash
      run: |
        echo TEST2 ${{ inputs.module }}.   # <-- ⛔️ This isn't echoing.  `inputs` seems to be clobbered by previous step.
```

### Relevant log output

```sh
Pipeline/test-orders] 🚀  Start image=nektos/act-environments-ubuntu:18.04
<snip>
| TEST1 some-module
[Pipeline/test-orders]     Success - Main echo TEST1 ${{ inputs.module }}

<snip>
[Pipeline/test-orders]  Runle Main actions/setup-node@v2
<snip>
[Pipeline/test-orders]     ##Success - Main actions/setup-/ode@v2
<snip>
| TEST2         <-- ⛔️ This should say "TEST2 some-module" instead!!!!
<snip>
[Pipeline/test-orders] 🏁  Job succeeded
```

### Additional information

I would expect it to echo "TEST2 some-module", but using actions/setup-node seems to clobber `inputs` so that `${{ inputs.module }}` is empty.

When I run it verbosely, after actions/setup-node, I no longer see `INPUT_MODULE` being set for subsequent steps.

**This works fine on GHA on GHE, which is why I think it might be an act issue.**

### PR DESCRIPTION
**Pull request by KnisterPeter**
This change fixes the composite action setup handling of inputs.

All inputs are taken from the env now. The env is composed of
the 'level above'.
For example:
- step env -> taken from run context
- action env -> taken from step env
- composite env -> taken from action env

Before this change the env setup for steps, actions and composite
run contexts was harder to understand as all parts looked into
one of these: parent run context, step, action, composite run context.

Now the 'data flow' is from higher levels to lower levels which should
make it more clean.

### ToDo before merging

- [x] add remote composite action test
- [ ] ~~add test for event inputs (e.g. `workflow_dispatch`)~~ Do not work in act yet. Here are some tests that show the support implemented: https://github.com/nektos/act/pull/1363
- [x] add proper support for `ActionRepository` and `ActionRef`

Fixes #1319

### CODE REVIEW

**Comment by KnisterPeter** [comment]
> My comment on the deactivated setup-node test step is a real issue.
> 
> Act or the workflow iteself have to santize the output of a workflow command `::add-matcher::` otherwise actions/runner fails the step, even if the test suite returns code 0.

Ah I see. I thought it was disabled, because the inputs where broken.
I'll add back the comment and skip that steps.

**Comment by KnisterPeter** [comment]
Now it's just the CodeCov upload

**Comment by cplee** [comment]
Looks great @KnisterPeter. Awesome improvement over readability!

File: pkg/runner/action.go [hunk 1]
```diff
@@ -193,6 +177,34 @@ func runActionImpl(step actionStep, actionDir string, remoteAction *remoteAction
...
+	} else {
+		rc.ActionRepository = ""
+		rc.ActionRef = ""
+	}
+	defer (func() {
+		// cleanup after the action is done, to avoid side-effects in
+		// the next step/action
+		rc.ActionRepository = ""
+		rc.ActionRef = ""
+	})()
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply pkg/runner/action.go:195]: Yes, that's a good point.
Obviously we don't have a test case for that. I will add one and as you suggested, will remove that fields from the `RunContext`. The bigger problem here is that it is taken from the `RunContext` to fill the `GithubContext`. That will make this change
a lot more complex.

Maybe I move it back and refactor that part in a next PR.

File: pkg/runner/action.go [hunk 2]
```diff
@@ -511,7 +499,6 @@ func runPreStep(step actionStep) common.Executor {
 			return rc.execJobContainer(containerArgs, *step.getEnv(), "", "")(ctx)
 
 		case model.ActionRunsUsingComposite:
-			step.getCompositeRunContext(ctx).updateCompositeRunContext(ctx, step.getRunContext(), step)
 			return step.getCompositeSteps().pre(ctx)
```
[the remark this answers is not in the document]
Comment (KnisterPeter) [reply pkg/runner/action.go:502]: I do look into this. And definitely add at least a simple 'remove composite action' test.

Create a helper to test local actions as remote actions would be a good idea, but not part of this PR.

**Comment by KnisterPeter** [comment]
> FYI You are removing `workflow_call` / `workflow_dispatch` `inputs` context support in this PR. I'm already using the `workflow_call` `inputs` context in github-act-runner, on an old version of act without the bug addressed in this PR.
> 
> Docu https://docs.github.com/en/actions/learn-github-actions/contexts#inputs-context
> 
> You are removing workflow_call / workflow_dispatch inputs context support in this PR. I'm already using the workflow_call inputs context in github-act-runner, on an old version of act without the bug addressed in this PR.

Another good catch. We do lack too many tests.
This is not dead code but required to work.

I will create a todo list in this PR which need to be addressed before we can merge this.
At least we should not introduce a lot more regressions.

**Comment by KnisterPeter** [comment]
> You are removing workflow_call / workflow_dispatch inputs context support in this PR

@ChristopherHX 
There was no mapping from inputs on the event to the `inputs` context, right?

Shouldn't be that hard to add but this is not a regression, is it?

**Comment by KnisterPeter** [comment]
> setupActionInputs

If we just remove `setupActionInputs` we break remote actions and it's inputs.

> Correct. That is a feature request for workflow_dispatch, but reusable workflows ( workflow_call ) doesn't have them in the github.event.

But you can probably create a patch where you place the inputs from reusable workflows into the `env` map and name them like action inputs (e.g. `INPUT_...`).
That would keep it consistent.

It would also be quite easy to implement that support as step type with the proper `env` map setup to support reusable workflows in act (which I have on my todo list for the future).

**Comment by KnisterPeter** [comment]
@ChristopherHX  I'm fine if you will provide a fix for #1319. We can abandon this and use your patch if you say it does fix the issue.

**Comment by KnisterPeter** [comment]
> I forget to mention, how to restore the boolean, number type if we are switching to env variables?
> In composite actions all input values are strings, but only there.

All inputs with actions are strings. Not just composite inputs.
The exception are `workflow_call` and `workflow_dispatch`. Both are currently not implemented in act.

**Comment by KnisterPeter** [comment]
> Undeclared inputs of composite actions disappeared, after this change. This might break other workflows, which ignore the warning.

Do you know if this is en-par with GitHub? If not I will change this again.

**Comment by KnisterPeter** [comment]
As stated [in the docs](https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_dispatchinputs) the `inputs` context should preserve the boolean type for `workflow_dispatch` inputs.

This need to be addressed in a different PR (e.g. #1363)

**Comment by KnisterPeter** [comment]
@cplee @ZauberNerd @catthehacker 
This fixes an issue which heavily breaks composite builds. Any review here would be helpful.

**Review by cplee — APPROVED** [review]

**Comment by cplee** [comment]
@KnisterPeter - approved. Should I trigger a new release after this merges?

**Pull request merged**